overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
 	return filePerms, nil
 }
 
-var allowablePathPatternRegexp = regexp.MustCompile(`^(/|(/[^/*{}]+)*(/\*|(/\*\*)?(/\*\.[^/*{}]+)?)?)$`)
+var (
+	// The following must be escaped as literals in a path pattern:
+	problematicChars = `\[\]`
+	// A single safe or escaped unsafe char in a path
+	safePathChar = fmt.Sprintf(`([^%s]|\\[%s])`, problematicChars, problematicChars)
+
+	// The following matches valid path patterns
+	allowablePathPattern       = fmt.Sprintf(`^/%s*$`, safePathChar)
+	allowablePathPatternRegexp = regexp.MustCompile(allowablePathPattern)
+)
+
+// Expands all groups in the given path pattern. Groups are enclosed by '{' '}'.
+// Returns a list of all the expanded path patterns, or an error if the given
+// pattern is invalid.
+func ExpandPathPattern(pattern string) ([]string, error) {
+	patternLen := len(pattern)
+	expanded := defaultPrefixes
+	currLiteralStart := 0
+	for i := 0; i < patternLen; i++ {
```

Comment:
This will iterate through by bytes rather than runes, is this okay? Perhaps `for i, c := range pattern {` could work? Or at least  `for i := range pattern { c := pattern[i]`.

Either way, storing `pattern[i]` would clean things up a little

Reply:
I think I need to iterate by runes, in case path patterns are non-ASCII. But I need a way to skip the next rune if it's preceded by a `'\\'` -- that's why I didn't iterate over `i, c`, and why I didn't define `c := pattern[i]`, since it might immediately be reassigned and I only use it 2 times after that anyway. I suppose a variable would do, but I still need a way of creating a substring starting after a particular rune index and continuing until another index.

Wouldn't `for i, c := range pattern` be essentially identical to `for i := range pattern`, at least in terms of byte vs rune iteration?

Reply:
Ah my mistake, I thought one iterated bytes but they both iterate runes